### DISCUSSION TITLE
0055 XLS-55d: Remit: Atomic Multi-Asset Payments for  XRPL Protocol Chains

### DISCUSSION THREAD

**Discussion by RichardAH** [XRPLF/XRPL-Standards#156] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by WietseWind** [comment]
Yay, payment + loyalty + voucher at once :)

Retail, store:
- Product price: €10
- Discount when sending 500 loyalty points: €1
- Voucher (URIToken) for 20% off

» Send one atomic TX with the voucher, 500 loyalty points & (10-1)*(1-0.2) Gatehub.EUR 🎉

**Reply by MarvinSunday** [reply]
I'm trying to understand the logic about the above transaction Wieste.
Is it the retail store that is sending the remit or is the product buyer?
I also understand that the Loyalty points will be minted as a new URIToken by the transaction, am I wrong?

**Reply by WietseWind** [reply]
It doesn't really matter who initiates, can be the store:

- Store: "that'll be 10€"
- Customer: "I have this voucher and I also would like to cash out 500 points"
- Store: "ok here's a sign request for a ttREMIT with (see exmaple)"

... can be the user:
- "OK, let me use this and that to check out"
- Click pay
- Get presented a sign request to pay accordingly

Loyalty points can be anything, I'd say simplest would just be an issued currency. But that's not the point of this proposal, it can be anything. The proposal is about atomic transactions with different representations of value (issued assets, native asset, URIToken, ...) where either all inner transactions (transfers of value) fail, or all succeed.

**Reply by MarvinSunday** [reply]
Crazy! I get it now Wieste, thanks this is awesome 😎
I feel like this transaction type ttRemit will help alot of real life applications to thrive on the Xahau.
Great one @RichardAH  👏🏾👏🏾

**Comment by tequdev** [comment]
It will be more processing compared to other transaction types, but is baseFee still the default?

**Reply by RichardAH** [reply]
It's actually less than a pathing payment (considerably less). It will be charged as a base fee unless someone can show objectively that it should be charged as more.

**Comment by mvadari** [comment]
Shouldn't the ability to create trustlines for someone else be opt-in instead of opt-out? I could imagine a scam where someone sends a bunch of remits with dust to every account that doesn't have `asfDisallowIncomingRemit` enabled/a hook preventing those transactions (likely most of them) just to bump up the number of trustlines their token has to make it seem more legit/game token rankings.

**Reply by RichardAH** [reply]
It's the type of spam people usually don't mind receiving because they can run a cleanup xapp/wallet tool and recover all the free reserves from all the spam tokens they didn't authorize

**Reply by mvadari** [reply]
I'm worried less about the spam and more about scams. Scammers can effectively now buy as many trustlines as they are willing to pay for to make their tokens seem legit and not like a rugpull (currently $0.05 per trustline, which is not much).

**Reply by RichardAH** [reply]
But don't you think after they've done that a couple times, the number of trustlines will no longer be a mark of trust?

**Reply by Silkjaer** [reply]
If you issue a token now, a lot of accounts will set trustlines anyway anticipating an airdrop.

Scammers can also create accounts and set trustlines themselves.

I think we're way past the point where the number of lines provide credibility. Even volume is easily faked.

**Reply by RichardAH** [reply]
Yeah Sybil attack is very easy here. If they have gas to burn just spin up 100,000 accounts and that's 100,000 trustlines for each of their tokens.

**Reply by mvadari** [reply]
Creating a new account is 5x as expensive as creating a trustline for an already-existing account. All those accounts that set up trustlines anyways can still opt-in to remits.

I'm not suggesting removing the feature of being able to pay for + create a trustline on behalf of someone else - I think that's a pretty neat feature - just that it should be opt-in instead of opt-out.

**Reply by hazardcookie** [reply]
I think the more technical minded folks might understand the amount of trustlines doesn't mean its credible. However i'd argue that the vast majority of end users value that metric. It's one of the metrics sites like [onthedex](https://onthedex.live/) and [xpmarkets](https://xpmarket.com/token/SOLO-rsoLo2S1kiGeCcn6hCUXVrCpGMWLrRrLZz) use as one of their key data points. Yes we might know that we shouldn't put much stock into that number, but if we put ourselves in the place of average users, i think they would find that to mean its "trustworthy".

**Reply by RichardAH** [reply]
XRPL allows push payments for XRP that can contain memos which contain advertisement for scams. Should we disallow push payments for XRP as well because they can be used by scammers to spam unsuspecting users? This is the logic, no?

**Reply by xVet** [reply]
Isn't that essentially how ETH and other evm chains work?

**Reply by Cbot-XRPL** [reply]
Im pretty certain this is how it works on evm.

**Comment by Mwni** [comment]
For this amendment to work, the XLS-35 `URITokens` amendment needs to be active as well. What happens if this amendment gets voted in, but URITokens don't?

**Reply by RichardAH** [reply]
You could always write an implementation of it that works with xls20?

**Comment by RichardAH** [comment]
Changed the title to clarify that this standard can apply to any XRPL protocol chain (although only one such chain with required xls-35 adoption currently exists).

**Comment by mvadari** [comment]
A couple of additional thoughts:
* Can there be a flag on `Remit` to allow senders to opt out of creating a trustline (if needed), if they don't want to give up their 2 XRP?
* What happens with authorized trustlines?
* What if someone sends a remit for `USD.bob` instead of `USD.Gatehub` (and some other currencies/URITokens) and creates that trustline? Wouldn't you still have to send back the other currencies and URITokens, defeating the purpose?

**Reply by WietseWind** [reply]
- If one opts out of creating a TL, and that was needed, the entire Remit would fail as the TX needing the TL would fail, thus all would fail.
- Authorized TL: if unauthorized: that TX would fail, so the entire Remit fails
- If someone sends the wrong thing, then yes, it needs to be reverted. But you can't prevent stupid ;)

**Reply by mvadari** [reply]
* Re flags: Yes, I'm saying that should be an option for senders, just like with authorized trustlines.
* Re sending something else: I'm not talking about mistakes, I'm talking about scams. Not re sending worthless tokens to many people, but tricking a merchant into thinking they sent a legitimate token when they actually sent something worthless.

**Reply by WietseWind** [reply]
> but tricking a merchant into thinking they sent a legitimate token when they actually sent something worthless

That's something they can catch on the receiving with a Hook, and simply by checking what comes in. They have to do and manage these checks anyway, because what if the client simply doesn't pay enough? 

Let's take a retail example, I'm in a store, have to pay 9USD+100 points. I pay 9 EUR. Store gives me my money back, points are never taken from my membership card. I either pay 9USD+points, or nothing at all. 

With Hooks this would be even nicer because the recipient kan simply reject anything that isn't USD, and again nothing would go through if it doesn't match those conditions.

**Comment by Cbot-XRPL** [comment]
Going along with previous comment. Would there need to be a flag if a sender did not want to fund creating the destination account? I could see a situation where the remit sender does not want to be the root of destination accounts.

**Comment by tequdev** [comment]
If defaultRipple is enabled for the destination account, wouldn't an attacker be able to steal funds via Rippling?

**Reply by dangell7** [reply]
Can you elaborate more on how you think this is possible?

**Reply by tequdev** [reply]
I checked the Remit Amendment, and it appears that the LimitAmount of the trustline automatically created by the Remit transaction is 0, so this does not seem to be a concern.

**Comment by tequdev** [comment]
If I send multiple types of tokens (native/issued), what value will meta.delivered_amount have?

**Reply by dangell7** [reply]
`delivered_amount` will be undefined because the `delivered_amount` is the `sfAmount` and in this case it doesn't exist on this transactor.

**Reply by tequdev** [reply]
There is a possibility of balance inconsistencies when depositing to the exchange via Partial Payment.

If the exchanges can set DisallowRemit flag, it's not a problem, but I don't know if all exchanges can handle it.

For reference:
https://github.com/XRPLF/XRPL-Standards/discussions/162#discussioncomment-8243889

**Reply by dangell7** [reply]
There is no Partial Payment with ttREMIT

**Comment by tequdev** [comment]
As others have mentioned, I prefer opt-in instead of opt-out.
Also, some kind of illegal token or URIToken should not be accepted by default.

Or it could happen that they send issued tokens to accounts on exchanges that have not set the disallow flag to put pressure on them to support tokens.

[a comment by an automated account is not shown]
